File: server/routes/customFields.js
const express = require('express');
const router = express.Router();
const db = require('../db/database');
const { v4: uuidv4 } = require('uuid');

// Helper to generate timestamp
const getTimestamp = () => new Date().toISOString();

// GET /api/custom-fields - List all custom fields (optional ?project_id= filter)
router.get('/', (req, res) => {
  try {
    const { project_id } = req.query;
    
    let query = `
      SELECT cf.*, p.name as project_name
      FROM custom_fields cf
      LEFT JOIN projects p ON cf.project_id = p.id
    `;
    let params = [];
    
    if (project_id) {
      // Get project-specific fields AND global fields (project_id IS NULL)
      query += ` WHERE cf.project_id = ? OR cf.project_id IS NULL`;
      params.push(project_id);
    }
    
    query += ` ORDER BY cf.sort_order ASC, cf.created_at ASC`;
    
    const fields = db.prepare(query).all(...params);
    
    // Parse options JSON for select/multiselect fields
    const parsedFields = fields.map(field => ({
      ...field,
      options: field.options ? JSON.parse(field.options) : null,
      required: !!field.required
    }));
    
    res.json({ success: true, data: parsedFields });
  } catch (error) {
    console.error('Error fetching custom fields:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'FETCH_ERROR', message: error.message } 
    });
  }
});

// GET /api/custom-fields/:id - Get single custom field
router.get('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const field = db.prepare(`
      SELECT cf.*, p.name as project_name
      FROM custom_fields cf
      LEFT JOIN projects p ON cf.project_id = p.id
      WHERE cf.id = ?
    `).get(id);
    
    if (!field) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Custom field not found' } 
      });
    }
    
    // Parse options JSON
    field.options = field.options ? JSON.parse(field.options) : null;
    field.required = !!field.required;
    
    res.json({ success: true, data: field });
  } catch (error) {
    console.error('Error fetching custom field:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'FETCH_ERROR', message: error.message } 
    });
  }
});

// POST /api/custom-fields - Create custom field
router.post('/', (req, res) => {
  try {
    const { name, field_type, project_id, options, required, sort_order } = req.body;
    
    // Validation
    if (!name || !field_type) {
      return res.status(400).json({ 
        success: false, 
        error: { code: 'VALIDATION_ERROR', message: 'Name and field_type are required' } 
      });
    }
    
    const validFieldTypes = ['text', 'number', 'date', 'select', 'multiselect', 'checkbox', 'url'];
    if (!validFieldTypes.includes(field_type)) {
      return res.status(400).json({ 
        success: false, 
        error: { code: 'VALIDATION_ERROR', message: `field_type must be one of: ${validFieldTypes.join(', ')}` } 
      });
    }
    
    // Validate options for select/multiselect fields
    if ((field_type === 'select' || field_type === 'multiselect') && !options) {
      return res.status(400).json({ 
        success: false, 
        error: { code: 'VALIDATION_ERROR', message: 'Options are required for select/multiselect fields' } 
      });
    }
    
    const id = uuidv4();
    const now = getTimestamp();
    const optionsJson = options ? JSON.stringify(options) : null;
    
    const stmt = db.prepare(`
      INSERT INTO custom_fields (id, name, field_type, project_id, options, required, sort_order, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);
    
    stmt.run(
      id, 
      name, 
      field_type, 
      project_id || null, 
      optionsJson, 
      required ? 1 : 0, 
      sort_order || 0,
      now, 
      now
    );
    
    const newField = db.prepare(`
      SELECT cf.*, p.name as project_name
      FROM custom_fields cf
      LEFT JOIN projects p ON cf.project_id = p.id
      WHERE cf.id = ?
    `).get(id);
    
    newField.options = newField.options ? JSON.parse(newField.options) : null;
    newField.required = !!newField.required;
    
    res.status(201).json({ success: true, data: newField });
  } catch (error) {
    console.error('Error creating custom field:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'CREATE_ERROR', message: error.message } 
    });
  }
});

// PUT /api/custom-fields/:id - Update custom field
router.put('/:id', (req, res) => {
  try {
    const { id } = req.params;
    const { name, field_type, project_id, options, required, sort_order } = req.body;
    
    // Check if field exists
    const existingField = db.prepare('SELECT * FROM custom_fields WHERE id = ?').get(id);
    if (!existingField) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Custom field not found' } 
      });
    }
    
    // Validate field_type if provided
    if (field_type) {
      const validFieldTypes = ['text', 'number', 'date', 'select', 'multiselect', 'checkbox', 'url'];
      if (!validFieldTypes.includes(field_type)) {
        return res.status(400).json({ 
          success: false, 
          error: { code: 'VALIDATION_ERROR', message: `field_type must be one of: ${validFieldTypes.join(', ')}` } 
        });
      }
    }
    
    const now = getTimestamp();
    const optionsJson = options !== undefined ? JSON.stringify(options) : existingField.options;
    
    const stmt = db.prepare(`
      UPDATE custom_fields 
      SET name = ?, field_type = ?, project_id = ?, options = ?, required = ?, sort_order = ?, updated_at = ?
      WHERE id = ?
    `);
    
    stmt.run(
      name ?? existingField.name,
      field_type ?? existingField.field_type,
      project_id !== undefined ? (project_id || null) : existingField.project_id,
      optionsJson,
      required !== undefined ? (required ? 1 : 0) : existingField.required,
      sort_order !== undefined ? sort_order : existingField.sort_order,
      now,
      id
    );
    
    const updatedField = db.prepare(`
      SELECT cf.*, p.name as project_name
      FROM custom_fields cf
      LEFT JOIN projects p ON cf.project_id = p.id
      WHERE cf.id = ?
    `).get(id);
    
    updatedField.options = updatedField.options ? JSON.parse(updatedField.options) : null;
    updatedField.required = !!updatedField.required;
    
    res.json({ success: true, data: updatedField });
  } catch (error) {
    console.error('Error updating custom field:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'UPDATE_ERROR', message: error.message } 
    });
  }
});

// DELETE /api/custom-fields/:id - Delete custom field
router.delete('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const existingField = db.prepare('SELECT * FROM custom_fields WHERE id = ?').get(id);
    if (!existingField) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Custom field not found' } 
      });
    }
    
    // Delete the field (cascade will delete associated values)
    db.prepare('DELETE FROM custom_fields WHERE id = ?').run(id);
    
    res.json({ success: true, data: { id, message: 'Custom field deleted successfully' } });
  } catch (error) {
    console.error('Error deleting custom field:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'DELETE_ERROR', message: error.message } 
    });
  }
});

module.exports = router;

File: server/internal/handlers/custom_fields.go
package handlers

import (
	"database/sql"
	"encoding/json"
	"net/http"

	"github.com/celestask/server/internal/db"
	"github.com/celestask/server/internal/middleware"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// Valid field types
var validFieldTypes = []string{"text", "number", "date", "select", "multiselect", "checkbox", "url"}

// CustomField represents a custom field entity
type CustomField struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	FieldType   string          `json:"field_type"`
	ProjectID   *int            `json:"project_id"`
	ProjectName *string         `json:"project_name"`
	Options     json.RawMessage `json:"options"`
	Required    bool            `json:"required"`
	SortOrder   int             `json:"sort_order"`
	CreatedAt   string          `json:"created_at"`
	UpdatedAt   string          `json:"updated_at"`
}

// CreateCustomFieldRequest represents the request body for creating a custom field
type CreateCustomFieldRequest struct {
	Name      string          `json:"name" binding:"required"`
	FieldType string          `json:"field_type" binding:"required"`
	ProjectID *int            `json:"project_id"`
	Options   json.RawMessage `json:"options"`
	Required  bool            `json:"required"`
	SortOrder int             `json:"sort_order"`
}

// UpdateCustomFieldRequest represents the request body for updating a custom field
type UpdateCustomFieldRequest struct {
	Name      *string         `json:"name"`
	FieldType *string         `json:"field_type"`
	ProjectID *int            `json:"project_id"`
	Options   json.RawMessage `json:"options"`
	Required  *bool           `json:"required"`
	SortOrder *int            `json:"sort_order"`
}

// isValidFieldType checks if the field type is valid
func isValidFieldType(fieldType string) bool {
	for _, ft := range validFieldTypes {
		if ft == fieldType {
			return true
		}
	}
	return false
}

// GetCustomFields handles GET /api/custom-fields
// Lists custom fields, optionally filtered by project_id
func GetCustomFields(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	projectID := c.Query("project_id")

	var rows *sql.Rows
	var err error

	if projectID != "" {
		// Get project-specific fields AND global fields (project_id IS NULL)
		rows, err = database.Query(`
			SELECT cf.id, cf.name, cf.field_type, cf.project_id, cf.options, cf.required, cf.sort_order, cf.created_at, cf.updated_at, p.name as project_name
			FROM custom_fields cf
			LEFT JOIN projects p ON cf.project_id = p.id
			WHERE cf.project_id = ? OR cf.project_id IS NULL
			ORDER BY cf.sort_order ASC, cf.created_at ASC
		`, projectID)
	} else {
		rows, err = database.Query(`
			SELECT cf.id, cf.name, cf.field_type, cf.project_id, cf.options, cf.required, cf.sort_order, cf.created_at, cf.updated_at, p.name as project_name
			FROM custom_fields cf
			LEFT JOIN projects p ON cf.project_id = p.id
			ORDER BY cf.sort_order ASC, cf.created_at ASC
		`)
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom fields"))
		return
	}
	defer rows.Close()

	fields := []CustomField{}
	for rows.Next() {
		var field CustomField
		var projectID sql.NullInt64
		var projectName sql.NullString
		var options []byte
		var required int

		err := rows.Scan(
			&field.ID,
			&field.Name,
			&field.FieldType,
			&projectID,
			&options,
			&required,
			&field.SortOrder,
			&field.CreatedAt,
			&field.UpdatedAt,
			&projectName,
		)
		if err != nil {
			c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom fields"))
			return
		}

		if projectID.Valid {
			id := int(projectID.Int64)
			field.ProjectID = &id
		}
		if projectName.Valid {
			field.ProjectName = &projectName.String
		}
		if options != nil {
			field.Options = options
		}
		field.Required = required == 1

		fields = append(fields, field)
	}

	if err := rows.Err(); err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom fields"))
		return
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(fields))
}

// GetCustomField handles GET /api/custom-fields/:id
// Gets a single custom field by ID
func GetCustomField(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	var field CustomField
	var projectID sql.NullInt64
	var projectName sql.NullString
	var options []byte
	var required int

	err := database.QueryRow(`
		SELECT cf.id, cf.name, cf.field_type, cf.project_id, cf.options, cf.required, cf.sort_order, cf.created_at, cf.updated_at, p.name as project_name
		FROM custom_fields cf
		LEFT JOIN projects p ON cf.project_id = p.id
		WHERE cf.id = ?
	`, id).Scan(
		&field.ID,
		&field.Name,
		&field.FieldType,
		&projectID,
		&options,
		&required,
		&field.SortOrder,
		&field.CreatedAt,
		&field.UpdatedAt,
		&projectName,
	)

	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Custom field"))
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom field"))
		return
	}

	if projectID.Valid {
		id := int(projectID.Int64)
		field.ProjectID = &id
	}
	if projectName.Valid {
		field.ProjectName = &projectName.String
	}
	if options != nil {
		field.Options = options
	}
	field.Required = required == 1

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(field))
}

// CreateCustomField handles POST /api/custom-fields
// Creates a new custom field
func CreateCustomField(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	var req CreateCustomFieldRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Name and field_type are required"))
		return
	}

	// Validate field type
	if !isValidFieldType(req.FieldType) {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("field_type must be one of: text, number, date, select, multiselect, checkbox, url"))
		return
	}

	// Validate options for select/multiselect fields
	if (req.FieldType == "select" || req.FieldType == "multiselect") && len(req.Options) == 0 {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Options are required for select/multiselect fields"))
		return
	}

	// Validate project_id if provided
	if req.ProjectID != nil && *req.ProjectID != 0 {
		var projExists bool
		if err := database.QueryRow("SELECT 1 FROM projects WHERE id = ?", *req.ProjectID).Scan(&projExists); err != nil {
			c.JSON(http.StatusBadRequest, middleware.NewValidationError("Project not found"))
			return
		}
	}

	id := uuid.New().String()
	sortOrder := req.SortOrder
	if sortOrder == 0 {
		sortOrder = 0
	}

	var options []byte
	if req.Options != nil {
		options = req.Options
	}

	_, err := database.Exec(`
		INSERT INTO custom_fields (id, name, field_type, project_id, options, required, sort_order, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))
	`,
		id,
		req.Name,
		req.FieldType,
		req.ProjectID,
		options,
		req.Required,
		sortOrder,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewCreateError("custom field"))
		return
	}

	// Fetch the created field
	var field CustomField
	var projectID sql.NullInt64
	var projectName sql.NullString
	var required int

	err = database.QueryRow(`
		SELECT cf.id, cf.name, cf.field_type, cf.project_id, cf.options, cf.required, cf.sort_order, cf.created_at, cf.updated_at, p.name as project_name
		FROM custom_fields cf
		LEFT JOIN projects p ON cf.project_id = p.id
		WHERE cf.id = ?
	`, id).Scan(
		&field.ID,
		&field.Name,
		&field.FieldType,
		&projectID,
		&options,
		&required,
		&field.SortOrder,
		&field.CreatedAt,
		&field.UpdatedAt,
		&projectName,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom field"))
		return
	}

	if projectID.Valid {
		id := int(projectID.Int64)
		field.ProjectID = &id
	}
	if projectName.Valid {
		field.ProjectName = &projectName.String
	}
	if options != nil {
		field.Options = options
	}
	field.Required = required == 1

	c.JSON(http.StatusCreated, middleware.NewSuccessResponse(field))
}

// UpdateCustomField handles PUT /api/custom-fields/:id
// Updates an existing custom field
func UpdateCustomField(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	// Check if field exists
	var existingField CustomField
	var existingProjectID sql.NullInt64
	var existingOptions []byte
	var existingRequired int

	err := database.QueryRow(`
		SELECT id, name, field_type, project_id, options, required, sort_order
		FROM custom_fields WHERE id = ?
	`, id).Scan(
		&existingField.ID,
		&existingField.Name,
		&existingField.FieldType,
		&existingProjectID,
		&existingOptions,
		&existingRequired,
		&existingField.SortOrder,
	)
	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Custom field"))
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom field"))
		return
	}

	var req UpdateCustomFieldRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Invalid request body"))
		return
	}

	// Validate field_type if provided
	if req.FieldType != nil && !isValidFieldType(*req.FieldType) {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("field_type must be one of: text, number, date, select, multiselect, checkbox, url"))
		return
	}

	// Build update query dynamically
	name := existingField.Name
	if req.Name != nil {
		name = *req.Name
	}

	fieldType := existingField.FieldType
	if req.FieldType != nil {
		fieldType = *req.FieldType
	}

	projectID := existingProjectID
	if req.ProjectID != nil {
		if *req.ProjectID == 0 {
			projectID = sql.NullInt64{Valid: false}
		} else {
			// Validate project exists
			var projExists bool
			if err := database.QueryRow("SELECT 1 FROM projects WHERE id = ?", *req.ProjectID).Scan(&projExists); err != nil {
				c.JSON(http.StatusBadRequest, middleware.NewValidationError("Project not found"))
				return
			}
			projectID = sql.NullInt64{Int64: int64(*req.ProjectID), Valid: true}
		}
	}

	options := existingOptions
	if req.Options != nil {
		options = req.Options
	}

	// Re-run select/multiselect invariant after computing the final field_type
	if (fieldType == "select" || fieldType == "multiselect") && len(options) == 0 {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Options are required for select/multiselect fields"))
		return
	}

	required := existingRequired
	if req.Required != nil {
		if *req.Required {
			required = 1
		} else {
			required = 0
		}
	}

	sortOrder := existingField.SortOrder
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	_, err = database.Exec(`
		UPDATE custom_fields 
		SET name = ?, field_type = ?, project_id = ?, options = ?, required = ?, sort_order = ?, updated_at = datetime('now')
		WHERE id = ?
	`,
		name,
		fieldType,
		projectID,
		options,
		required,
		sortOrder,
		id,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewUpdateError("custom field"))
		return
	}

	// Fetch the updated field
	var field CustomField
	var newProjectID sql.NullInt64
	var projectName sql.NullString
	var newOptions []byte
	var newRequired int

	err = database.QueryRow(`
		SELECT cf.id, cf.name, cf.field_type, cf.project_id, cf.options, cf.required, cf.sort_order, cf.created_at, cf.updated_at, p.name as project_name
		FROM custom_fields cf
		LEFT JOIN projects p ON cf.project_id = p.id
		WHERE cf.id = ?
	`, id).Scan(
		&field.ID,
		&field.Name,
		&field.FieldType,
		&newProjectID,
		&newOptions,
		&newRequired,
		&field.SortOrder,
		&field.CreatedAt,
		&field.UpdatedAt,
		&projectName,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom field"))
		return
	}

	if newProjectID.Valid {
		id := int(newProjectID.Int64)
		field.ProjectID = &id
	}
	if projectName.Valid {
		field.ProjectName = &projectName.String
	}
	if newOptions != nil {
		field.Options = newOptions
	}
	field.Required = newRequired == 1

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(field))
}

// DeleteCustomField handles DELETE /api/custom-fields/:id
// Deletes a custom field
func DeleteCustomField(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	// Check if field exists
	var existingField CustomField
	err := database.QueryRow(`
		SELECT id FROM custom_fields WHERE id = ?
	`, id).Scan(&existingField.ID)
	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Custom field"))
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("custom field"))
		return
	}

	// Delete the field (cascade will delete associated values)
	_, err = database.Exec(`DELETE FROM custom_fields WHERE id = ?`, id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewDeleteError("custom field"))
		return
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(gin.H{
		"id":      id,
		"message": "Custom field deleted successfully",
	}))
}
